rebuild/src/secret_key.cc

```diff
@@ -0,0 +1,171 @@
+#include "secret_key.h"
+
+void SecretKey::Init(
+    Napi::Env env, Napi::Object &exports, BlstTsAddon *module) {
+    Napi::HandleScope scope(env);
+    auto proto = {
+        StaticMethod(
+            "fromKeygen",
+            &SecretKey::FromKeygen,
+            static_cast<napi_property_attributes>(
+                napi_static | napi_enumerable)),
+        StaticMethod(
+            "deserialize",
+            &SecretKey::Deserialize,
+            static_cast<napi_property_attributes>(
+                napi_static | napi_enumerable)),
+        InstanceMethod(
+            "serialize",
+            &SecretKey::Serialize,
+            static_cast<napi_property_attributes>(napi_enumerable)),
+        InstanceMethod(
+            "toPublicKey",
+            &SecretKey::ToPublicKey,
+            static_cast<napi_property_attributes>(napi_enumerable)),
+        InstanceMethod(
+            "sign",
+            &SecretKey::Sign,
+            static_cast<napi_property_attributes>(napi_enumerable)),
+    };
+
+    Napi::Function ctr = DefineClass(env, "SecretKey", proto, module);
+    module->_secret_key_ctr = Napi::Persistent(ctr);
+    exports.Set(Napi::String::New(env, "SecretKey"), ctr);
+
+    exports.Get("BLST_CONSTANTS")
+        .As<Napi::Object>()
+        .Set(
+            Napi::String::New(env, "SECRET_KEY_LENGTH"),
+            Napi::Number::New(env, BLST_TS_SECRET_KEY_LENGTH));
+}
+
+Napi::Value SecretKey::FromKeygen(const Napi::CallbackInfo &info) {
+    BLST_TS_FUNCTION_PREAMBLE(info, env, module)
+    Napi::Value ikm_value = info[0];
+
+    BLST_TS_UNWRAP_UINT_8_ARRAY(ikm_value, ikm, "ikm")
+    // Check for less than 32 bytes so consumers don't accidentally create
+    // zero keys 
+    if (ikm.ByteLength() < BLST_TS_SECRET_KEY_LENGTH) {
+        std::ostringstream msg;
+        msg << "ikm must be greater than or equal to "
+            << BLST_TS_SECRET_KEY_LENGTH << " bytes";
+        Napi::TypeError::New(env, msg.str()).ThrowAsJavaScriptException();
+        return env.Undefined();
+    }
+
+    bool is_external = true;
+    Napi::Object wrapped = module->_secret_key_ctr.New(
+        {Napi::External<bool>::New(env, &is_external)});
+    SecretKey *sk = SecretKey::Unwrap(wrapped);
```

Comment:
is it guaranteed not to return a nullptr or complete garbage?

Reply:
Yes. If the isolate does not return an object correctly, nor the correct type of object to get unwrapped by the overloaded class method it throws an error.  I am catching them everywhere wrapped objects are being passed into the functions from user land but in this instance it is created immediately before and the isolate is atomic so it will be guaranteed to exist.  Or a hard OOM error will have occurred on the line before but that is far more serious and will have stopped the world (kernel panic) anyway.